Doc: Add UITester documentation in User Manual (#1263)
enthought/traitsui: docs/source/traitsui_user_manual/testing.rst

            ),
            id="person_item",    # <--- add this
        )
    )

    obj = App()
    tester = UITester()
    with tester.create_ui(obj, dict(view=view)) as ui:
        wrapper = tester.find_by_id(ui, "person_item")

The returned value ``wrapper`` will be used in the following steps.

Perform a user interaction or continue navigation
-------------------------------------------------
The returned value of |UITester.find_by_name| and |UITester.find_by_id|
is an instance of |UIWrapper| on which you may either navigate into more
nested GUI elements, or perform a user interaction on the GUI element.

The supported actions can be found via the |UIWrapper.help| method::

    from traitsui.testing.api import UITester

    class Person(HasTraits):
        name = Str()

    class App(HasTraits):
        title = Str()
        person = Instance(Person, ())

    view = View(
        Item(name="title"),
        Item(name="person", editor=InstanceEditor(), style="custom"))

    obj = App()
    tester = UITester()
    with tester.create_ui(obj, dict(view=view)) as ui:
        title_field = tester.find_by_name(ui, "title")

>         # Print some help message

Reviewer: ```suggestion
        # Print information about the interactions and navigation possible on this object.
```

The "Locations" section shows the types of objects that can be used with
|UIWrapper.locate|. They are objects that allow developers to navigate
further into the the current GUI element.

In this example, ``title_field`` is wrapping a textbox, no further
nested GUI elements can be located and therefore there are no locations
supported.

Most of the time these objects can be imported from
``traitsui.testing.api``.

Perform a user interaction to modify GUI state
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
To modify the GUI state, we use |UIWrapper.perform| with an object whose
type is supported (as shown in the "Interactions" section from the help
message).

Say we want to modify the value in the text box as if the user has typed
in it, we can use the ``KeySequence`` object together with
|UIWrapper.perform|::

    from traitsui.testing.api import UITester, KeySequence

    obj = App()
    tester = UITester()
    with tester.create_ui(obj, dict(view=view)) as ui:
        title_field = tester.find_by_name(ui, "title")
        title_field.perform(KeySequence("New Title"))
        assert obj.title == "New Title"

We can then check the trait being edited is updated (many times,

> once per keystroke).

Reviewer: I think the "many times, once per keystroke" deserves more explanation. It essentialy means that `enter_set` is `False` so the trait is updated with every key press. Is that correct?

Modify trait and inspect GUI state
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
To check the GUI state is what we expect, we use |UIWrapper.inspect|.

In the above example, ``title_field`` wrapping a text box also supports a

> interaction type called ``DisplayedText``, so we can use that to check when the

Reviewer: ```suggestion
interaction type called ``DisplayedText``. We can use that to check that when the
```

    Interactions
    ------------
    No interactions are supported.

    Locations
    ---------
    <class 'traitsui.testing.tester.locator.TargetById'>
        A locator for locating the next UI target using an id.

        Attributes
        ----------
        id : str

    <class 'traitsui.testing.tester.locator.TargetByName'>
        A locator for locating the next UI target using a name.

        Attributes
        ----------
        name : str

We can see that |TargetById| and |TargetByName| are available for
locating the next GUI element. We can locate the text box for
``person.name`` using ``TargetByName``::

    from traitsui.testing.api import UITester, TargetByName

    obj = App()
    tester = UITester()
    with tester.create_ui(obj, dict(view=view)) as ui:
        person_pane = tester.find_by_name(ui, "person")
        person_name_field = person_pane.locate(TargetByName("name"))

In fact, |UIWrapper.find_by_id| and |UIWrapper.find_by_name| are simply
aliases for |TargetById| and |TargetByName| respectively, so you
can also write::

    person_name_field = person_pane.find_by_name("name")

The returned value is again an instance of |UIWrapper|, so we can repeat

> steps 2 oe 3.

Reviewer: ```suggestion
steps 2 or 3.
```

    person_pane.locate(...).locate(...).locate(...).perform(...)

Examples
========

Several test examples can be found for testing
:ref:`TraitsUI\'s own demos<traitsui-demo>`.

Editors
-------
- :github-demo:`ButtonEditor <Standard_Editors/tests/test_ButtonEditor_demo.py>`
- :github-demo:`ButtonEditor <Standard_Editors/tests/test_ButtonEditor_simple_demo.py>`
- :github-demo:`CheckListEditor <Standard_Editors/tests/test_CheckListEditor_simple_demo.py>`
- :github-demo:`EnumEditor <Standard_Editors/tests/test_EnumEditor_demo.py>`
- :github-demo:`InstanceEditor <Standard_Editors/tests/test_InstanceEditor_demo.py>`
- :github-demo:`ListEditor <Advanced/tests/test_List_editor_notebook_selection_demo.py>`
- :github-demo:`TextEditor <Standard_Editors/tests/test_TextEditor_demo.py>`

Applications
------------
- :github-demo:`Converter <Applications/tests/test_converter.py>`

Debugging
=========
When you want to sanity check the test is doing what you intended, or when
you want to debug an issue, the ``delay`` parameter may be useful for
slowing down the test so that you can see the GUI being updated.

Example::

    from traitsui.testing.api import UITester, KeySequence

    obj = App()

>     tester = UITester(delay=50)    # delay is set

Reviewer: i haven't looked at the docs for the `UITester` but i guess 50 is milliseconds.

Note that there are limitations as to how truthful the GUI looks visually
compared to manual testing.

FAQ
===

.. rubric:: Is UITester GUI toolkit specific?

It depends. The API of UITester is toolkit independent. As long as the behavior
being tested is also toolkit independent, |UITester| should support test code
with no toolkit specific logic. However if the underlying GUI behavior
varies from toolkit to toolkit, the test code using |UITester| will need to
adjust for that.

.. rubric:: Which testing framework should I use with UITester?

|UITester| does not depend on any testing framework. You should be able to use
it with any testing framework (e.g. unittest, pytest).

.. rubric:: Is UITester compatible with PyFace ModalDialogTester?

Yes. For example, you can use |UITester| to launch a modal dialog, and use
ModelDialogTester to close it.

.. rubric:: Is UITester compatible with PyFace GuiTestAssistant?

Yes.

|UIWrapper.perform| and |UIWrapper.inspect| automatically request GUI
events to be processed. Where they are used entirely for modifying GUI states,
you may find uses of some of GuiTestAssistant features no longer necessary.

> .. rubric:: I use the delay parameter to play back my test but it does not look identical to when I test the GUI manually, why?

Reviewer: is it not possible to wrap `.. rubric::` lines?

Author: It is possible actually. Fixing...

   code are not maintained together, tests may be subject to breakages caused
   by internal changes of the UI editors. With that, projects are encouraged
   to contribute testing support upstream for testing UI editors they do not
   maintain.

Terminology
-----------

Before we start, we need to define some terminology:

* Target
    is an object on which we can perform an action (e.g. a mouse click) to
    modify an application state, or retrieve a GUI application state, or
    search for other contained targets (e.g. a table widget that contains
    buttons and text).

    An instance of |UIWrapper| wraps a target under the attribute
    ``_target``.

* Interaction
    is an object that wraps the information for performing an action
    or retrieving GUI state(s), but it does not necessarily contain information
    of a Target. For example, both |MouseClick| and |DisplayedText| are
    interactions that can be used against different targets. An interaction can
    also be specialized for a specific target if needed.

    |UIWrapper.perform| and |UIWrapper.inspect| handle an interaction.

* Location
    is an object that wraps the information for searching a target from a

>     container target, but it does not necessary contains information specific
>     of a Target. For example, both |TargetById| and |TargetByName| are

Reviewer: ```suggestion
    container target, but it does not necessarily contain information specific
    to a Target. For example, both |TargetById| and |TargetByName| are
```

a |TargetRegistry|.

For the purpose of this document, suppose we want to perform many mouse clicks
on a UI component, but instead of calling ``perform(MouseClick())`` many times
in a loop like this::

    my_widget = UITester().find_by_id(ui, "some_id")
    for _ in range(10):
        my_widget.perform(MouseClick())

We want to exercise the mouse click many times by invoking |UIWrapper.perform| once
only::

    my_widget = UITester().find_by_id(ui, "some_id")
    my_widget.perform(ManyMouseClick(n_times=10))

First, we need to define this ``ManyMouseClick`` object::

    class ManyMouseClick:
        def __init__(self, n_times):
            self.n_times = n_times

Next, we need to know which object implements the GUI component. This is where
implementation details start to come in. We can inspect the object being
wrapped::

    >>> my_widget
    <traitsui.testing.tester.ui_wrapper.UIWrapper object at 0x7f940a3f10b8>
    >>> my_widget._target
    <package.ui.qt.shiny_button.ShinyButton object at 0x7fc90fb3b570>

The target is an instance of a ``ShinyButton`` class (made up
for this document). In this object, there is an instance of Qt QPushButton

> widget on which we want the mouse click to occur.

Reviewer: ```suggestion
widget which we want click with the mouse.
```

The returned value is the returned value from ``many_mouse_click``, which is
``None``.

Add support for locating a nested GUI element
---------------------------------------------

Support for |UIWrapper.locate| can be extended by registering additional
location type and resolution logic via |TargetRegistry.register_solver| on
a |TargetRegistry|.

Suppose we have a custom UI editor that contains some buttons. The objective of
a test is to click a specific button with a given label. We will therefore need
to locate the button with the given label before a mouse click can be
performed.

The test code we wanted to achieve looks like this::

    container = UITester().find_by_id(ui, "some_container")
    button_wrapper = container.locate(NamedButton("OK"))

where the targets are::

    >>> container._target
    <package.ui.qt.shiny_panel.ShinyPanel object at 0x7f940a3f10b8>
    >>> button_wrapper._target
    <package.ui.qt.shiny_button.ShinyButton object at 0x7fbcc3a63438>

We can define the new ``NamedButton`` location type::

    class NamedButton:
        ''' Locator for locating a push button by label.'''
        def __init__(self, label):
            self.label = label

> Say ``ShinyPanel`` is keeping track of the buttons by names with a
> dictionary called ``_buttons``. Then the logic to retrieving a button from a

Reviewer: ```suggestion
Say ``ShinyPanel`` keeps track of the buttons with a dictionary called
``_buttons`` where the names of the buttons are the keys of the dictionary.
Then the logic to retrieving a button from a
```

.. _testing:

============================

> Testing TraitsUI Application

Reviewer: nit : "Testing a TraitsUI Application" or "Testing TraitsUI Applications"

testing, then such visual discrepancies may have to be tolerated given there
are toolkit-dependent and platform-dependent limitations as to what can be
achieved for programmatically imitating user interactions.

.. _advanced-testing:

Advanced Testing
================

|UITester| supports testing TraitsUI editors with various user interaction
logic. However it is possible that projects' test code may require
additional logic that is not supported by |UITester| by default. Furthermore,
some projects may implement and maintain their own custom UI editors. Those
custom UI editors are also by default not supported by |UITester|.

The API allows extension such that

* projects can test TraitsUI editors with user interactions that do not
  come supported by default.
* projects can reuse the testing API for testing custom editors.

.. note::
   Extending support for testing a UI editor often requires knowledge of the
   implementation details of the editor. If UI editor and the testing support
   code are not maintained together, tests may be subject to breakages caused
   by internal changes of the UI editors. With that, projects are encouraged

>    to contribute testing support upstream for testing UI editors they do not

Author: I think I was using the word upstream to mean something close to the context mentioned in this [article](https://en.wikipedia.org/wiki/Upstream_(software_development)). So in the context of an external project using TraitsUI, this sentence encourages developers of these projects to contribute test support extension back to TraitsUI so that it can be maintained here, and thereby reducing their tests' exposure to implementation details.

Maybe I should just remove this sentence. Contributing patches and codes upstream when you can is generally a good thing to do, so there is no need to repeat it here.

>    to contribute testing support upstream for testing UI editors they do not

Reviewer 2: That makes sense, I think it is perfectly fine if you want to leave it.  Its up to you

to locate the button with the given label before a mouse click can be
performed.

The test code we wanted to achieve looks like this::

    container = UITester().find_by_id(ui, "some_container")
    button_wrapper = container.locate(NamedButton("OK"))

where the targets are::

    >>> container._target
    <package.ui.qt.shiny_panel.ShinyPanel object at 0x7f940a3f10b8>
    >>> button_wrapper._target
    <package.ui.qt.shiny_button.ShinyButton object at 0x7fbcc3a63438>

We can define the new ``NamedButton`` location type::

    class NamedButton:
        ''' Locator for locating a push button by label.'''
        def __init__(self, label):
            self.label = label

Say ``ShinyPanel`` keeps track of the buttons with a dictionary called
``_buttons`` where the names of the buttons are the keys of the dictionary.
Then the logic to retrieving a button from a
label can be written like this::

    def get_button(wrapper, location):
        """ Returns a ShinyButton from a UIWrapper wrapping ShinyPanel."""
        # wrapper is an instance of UIWrapper
        # location is an instance of NamedButton
        return wrapper.target._buttons[location.label]

> 

Reviewer 2: should we mention that this signature is require by register_solver, like you did explicitly for register_handler in the example above?  It is probably clear by reading it, but it might not hurt to say it explicitly

The solvers can then be registered for the container UI target::

    registry = TargetRegistry()
    registry.register_solver(
        target_class=ShinyPanel,
        locator_class=NamedButton,
        solver=get_button,
    )

> Similar to |TargetRegistry.register_solver|, by setting the ``target_class``

Reviewer 2: should this be similar to register_handler?
